### ISSUE TITLE
Frequent leader elections / multiple route connections created / cluster node becomes unresponsive

### ISSUE THREAD

**Issue by anXieTyPB** [nats-io/nats-streaming-server#629]
Nats-streaming-server version 0.10.2 running in a customized docker container running in AWS ECS. We managed to recreate the problem on a local workstation that uses Docker with HyperV integration.

We have stability issues running nats streaming in cluster mode. Our cluster consists of 3 docker containers. As soon as a leader is reelected, clients appear to have problems using their nats connections. Some nats streaming server instances even seem to refuse connections for a limited timeframe until either recovering or becoming unresponsive overall. The only current solution is to shut down the container and create a new container whose nats instance rejoins the existing cluster, effectively replacing the frozen nats instance.

We are currently using the following configurations for our cluster, consisting of "a", "b" and "c". The configuration shown resembles the workstation test environment. In addition, the test setup consists of one producer that produces up to ~120 messages per 10 seconds in 1 topic. One subscriber listens for messages on that topic at the same time.

```
# NATS specific configuration
port: 4222
http_port: 8222
cluster {
  listen: 0.0.0.0:6222
  routes: [
    "nats://hostip:6223",
    "nats://hostip:6224"
  ]
}  

# NATS Streaming specific configuration
streaming {
  store: file
  dir: /mnt/natsdata/a/store
  sd: true
  cluster {
    node_id: "a"
    peers: ["b", "c"]
    log_path: /mnt/natsdata/a/log
  }
  store_limits: {
    max_channels: 1024
    max_msgs: 1000000
    max_bytes: 2GB
    max_age: "336h"
  }
}



# NATS specific configuration
port: 4223
http_port: 8223
cluster {
  listen: 0.0.0.0:6223
  routes: [
    "nats://hostip:6222",
    "nats://hostip:6224"
  ]
}

# NATS Streaming specific configuration
streaming {
  store: file
  dir: /mnt/natsdata/b/store
  sd: true
  cluster {
    node_id: "b"
    peers: ["a", "c"]
    log_path: /mnt/natsdata/b/log
  }
  store_limits: {
    max_channels: 1024
    max_msgs: 1000000
    max_bytes: 2GB
    max_age: "336h"
  }
}

# NATS specific configuration
port: 4224
http_port: 8224
cluster {
  listen: 0.0.0.0:6224
  routes: [
    "nats://hostip:6222",
    "nats://hostip:6223"
  ]
}

# NATS Streaming specific configuration
streaming {
  store: file
  dir: /mnt/natsdata/c/store
  sd: true
  cluster {
    node_id: "c"
    peers: ["a", "b"]
    log_path: /mnt/natsdata/c/log
  }
  store_limits: {
    max_channels: 1024
    max_msgs: 1000000
    max_bytes: 2GB
    max_age: "336h"
  }
}
```

We then use above configurations to build three different docker containers using the following dockerfile:

```
FROM alpine:3.7

LABEL NATS Streaming Customized

ARG MAIN_PORT=4222
ARG CLUSTER_PORT=6222
ARG HTTP_PORT=8222
ARG CONFIG_FILE

ENV NATSS_VERSION=v0.10.2

EXPOSE ${MAIN_PORT} ${CLUSTER_PORT} ${HTTP_PORT}

RUN apk update && \
    apk add --no-cache ca-certificates openssl

RUN wget -q https://github.com/nats-io/nats-streaming-server/releases/download/${NATSS_VERSION}/nats-streaming-server-${NATSS_VERSION}-linux-amd64.zip -O /tmp/nats-streaming-server.zip && \
    unzip /tmp/nats-streaming-server.zip -d / && \
    mv /nats-streaming-server-${NATSS_VERSION}-linux-amd64 /opt/ && \
    rm /tmp/nats-streaming-server.zip

RUN apk add --no-cache bash

COPY ${CONFIG_FILE} /opt/config.cfg

CMD ["/opt/nats-streaming-server", "-sc", "/opt/config.cfg"]

```

As follows is log output that made us curious about whether everything is working as intended. The host IP in the output was replaced with hostIP. The three nats instance connect to each other via hostip:6222, hostip:6223 and hostip:6224.

First thing that was striking is the count of leadership elections while running:
```
[1] 2018/07/18 09:29:20.330066 [INF] STREAM: server became leader, performing leader promotion actions
[1] 2018/07/18 09:29:20.379751 [INF] STREAM: finished leader promotion actions
[1] 2018/07/18 09:29:21.578353 [INF] STREAM: server lost leadership, performing leader stepdown actions
[1] 2018/07/18 09:29:21.737896 [INF] STREAM: finished leader stepdown actions
[1] 2018/07/18 09:29:23.454058 [INF] STREAM: server became leader, performing leader promotion actions
[1] 2018/07/18 09:29:23.954671 [INF] STREAM: finished leader promotion actions
[1] 2018/07/18 09:29:23.954688 [ERR] STREAM: Error on leadership acquired: leadership lost while committing log
[1] 2018/07/18 09:29:23.954693 [INF] STREAM: server lost leadership, performing leader stepdown actions
[1] 2018/07/18 09:29:24.757142 [INF] STREAM: finished leader stepdown actions
[1] 2018/07/18 09:29:25.779429 [INF] STREAM: server became leader, performing leader promotion actions
[1] 2018/07/18 09:29:26.274014 [INF] STREAM: finished leader promotion actions
[1] 2018/07/18 09:29:26.274028 [ERR] STREAM: Error on leadership acquired: leadership lost while committing log
[1] 2018/07/18 09:29:26.274033 [INF] STREAM: server lost leadership, performing leader stepdown actions
[1] 2018/07/18 09:29:27.991717 [INF] STREAM: finished leader stepdown actions
[1] 2018/07/18 09:29:30.789739 [INF] STREAM: server became leader, performing leader promotion actions
[1] 2018/07/18 09:29:31.289910 [INF] STREAM: finished leader promotion actions
[1] 2018/07/18 09:29:31.289938 [ERR] STREAM: Error on leadership acquired: leadership lost while committing log
[1] 2018/07/18 09:29:31.289944 [INF] STREAM: server lost leadership, performing leader stepdown actions
[1] 2018/07/18 09:29:31.724181 [INF] STREAM: finished leader stepdown actions
[1] 2018/07/18 09:29:32.483502 [INF] STREAM: server became leader, performing leader promotion actions
[1] 2018/07/18 09:29:33.064888 [INF] STREAM: finished leader promotion actions
[1] 2018/07/18 09:29:33.064901 [ERR] STREAM: Error on leadership acquired: leadership lost while committing log
[1] 2018/07/18 09:29:33.064904 [INF] STREAM: server lost leadership, performing leader stepdown actions
[1] 2018/07/18 09:29:33.408147 [INF] STREAM: finished leader stepdown actions
[1] 2018/07/18 09:29:34.332365 [INF] STREAM: server became leader, performing leader promotion actions
[1] 2018/07/18 09:29:34.832570 [INF] STREAM: finished leader promotion actions
[1] 2018/07/18 09:29:34.832658 [ERR] STREAM: Error on leadership acquired: leadership lost while committing log
[1] 2018/07/18 09:29:34.832679 [INF] STREAM: server lost leadership, performing leader stepdown actions
[1] 2018/07/18 09:29:35.234934 [INF] STREAM: finished leader stepdown actions
[1] 2018/07/18 09:29:36.446047 [INF] STREAM: Deleting raft logs from 20057 to 20058
[1] 2018/07/18 09:29:36.446191 [INF] STREAM: Deletion took 142.6µs

...

[1] 2018/07/18 09:47:25.611127 [INF] STREAM: server lost leadership, performing leader stepdown actions
[1] 2018/07/18 09:47:27.091716 [INF] STREAM: finished leader stepdown actions
[1] 2018/07/18 09:47:27.497898 [INF] STREAM: server became leader, performing leader promotion actions
[1] 2018/07/18 09:47:27.594223 [INF] STREAM: finished leader promotion actions

...

[1] 2018/07/18 09:03:45.175753 [DBG] STREAM: [Client:781f8190-8a69-11e8-b458-f94730128e42] Added member to durable queue subscription, subject=PRODUCT.ERP, inbox=_INBOX.AFDCFN2DDM38CH7BN5OADD, queue=AQueue Consumer:MyConsumer, subid=4
[1] 2018/07/18 09:03:51.752622 [INF] STREAM: server became leader, performing leader promotion actions
[1] 2018/07/18 09:03:51.821795 [INF] STREAM: finished leader promotion actions
[1] 2018/07/18 09:03:52.741442 [INF] STREAM: server lost leadership, performing leader stepdown actions
[1] 2018/07/18 09:03:53.088813 [INF] STREAM: finished leader stepdown actions
[1] 2018/07/18 09:03:54.350080 [INF] STREAM: server became leader, performing leader promotion actions
[1] 2018/07/18 09:03:55.192489 [INF] STREAM: finished leader promotion actions
[1] 2018/07/18 09:03:56.821066 [INF] STREAM: server lost leadership, performing leader stepdown actions
[1] 2018/07/18 09:03:58.945087 [INF] STREAM: finished leader stepdown actions
[1] 2018/07/18 09:03:58.945167 [INF] STREAM: server became leader, performing leader promotion actions
[1] 2018/07/18 09:03:59.385226 [INF] STREAM: finished leader promotion actions
[1] 2018/07/18 09:03:59.385303 [ERR] STREAM: Error on leadership acquired: leadership lost while committing log
[1] 2018/07/18 09:03:59.385309 [INF] STREAM: server lost leadership, performing leader stepdown actions
[1] 2018/07/18 09:03:59.661667 [INF] STREAM: finished leader stepdown actions
[1] 2018/07/18 09:04:00.926032 [INF] STREAM: server became leader, performing leader promotion actions
[1] 2018/07/18 09:04:01.000600 [INF] STREAM: finished leader promotion actions
[1] 2018/07/18 09:04:04.225938 [INF] STREAM: server lost leadership, performing leader stepdown actions
[1] 2018/07/18 09:04:04.255029 [INF] STREAM: finished leader stepdown actions
[1] 2018/07/18 09:04:05.569304 [INF] STREAM: server became leader, performing leader promotion actions
[1] 2018/07/18 09:04:06.069551 [INF] STREAM: finished leader promotion actions
[1] 2018/07/18 09:04:06.069661 [ERR] STREAM: Error on leadership acquired: leadership lost while committing log
[1] 2018/07/18 09:04:06.069735 [INF] STREAM: server lost leadership, performing leader stepdown actions
[1] 2018/07/18 09:04:06.104415 [INF] STREAM: finished leader stepdown actions
[1] 2018/07/18 09:04:08.302767 [INF] STREAM: Deleting raft logs from 7843 to 7845
[1] 2018/07/18 09:04:08.302988 [INF] STREAM: Deletion took 218.2µs
```

Second thing is after leader election concludes, lots of routes are created to the same cluster node.
```
[1] 2018/07/18 09:29:54.425968 [INF] hostip:6223 - rid:13 - Route connection created
[1] 2018/07/18 09:29:55.464313 [INF] hostip:6223 - rid:14 - Route connection created
[1] 2018/07/18 09:29:56.537748 [INF] hostip:6223 - rid:15 - Route connection created
[1] 2018/07/18 09:29:57.657936 [INF] hostip:6223 - rid:16 - Route connection created
[1] 2018/07/18 09:29:57.816937 [INF] hostip:6223 - rid:17 - Route connection created
[1] 2018/07/18 09:29:57.915439 [INF] hostip:6223 - rid:18 - Route connection created
[1] 2018/07/18 09:29:58.072570 [INF] hostip:6223 - rid:19 - Route connection created
[1] 2018/07/18 09:29:58.163819 [INF] hostip:6223 - rid:20 - Route connection created
[1] 2018/07/18 09:29:58.275567 [INF] hostip:6223 - rid:21 - Route connection created
[1] 2018/07/18 09:29:58.454568 [INF] hostip:6223 - rid:22 - Route connection created
[1] 2018/07/18 09:29:58.583304 [INF] hostip:6223 - rid:23 - Route connection created
...
[1] 2018/07/18 09:30:01.380984 [INF] hostip:6223 - rid:111 - Route connection created
[1] 2018/07/18 09:30:01.413447 [INF] hostip:6223 - rid:113 - Route connection created
[1] 2018/07/18 09:30:01.413521 [INF] hostip:6223 - rid:112 - Route connection created
[1] 2018/07/18 09:30:01.436997 [INF] hostip:6223 - rid:114 - Route connection created
[1] 2018/07/18 09:30:01.446385 [INF] hostip:6223 - rid:115 - Route connection created
[1] 2018/07/18 09:30:01.457837 [INF] hostip:6223 - rid:116 - Route connection created
[1] 2018/07/18 09:30:01.458015 [INF] hostip:6223 - rid:117 - Route connection created
[1] 2018/07/18 09:30:01.466721 [INF] hostip:6223 - rid:118 - Route connection created

```

Can you provide information on whether the behaviour shown in the logs is intended behaviour?
If yes, do you have recommendations on how to improve the setup?

**Comment by kozlovic** [comment]
I am not sure what's going on, but the fact that you see so many routes created really indicate that there is a network communication issues between nodes. If it was only the leader election we could suspect an issue with RAFT or something, but again, why so many routes created other than because it gets disconnected right away?
(note that the `Route connection created` is displayed as INF, but unfortunately, the route connection closed is in DBG, this is why you don't see them. We will address that in gnatsd.)

**Comment by kozlovic** [comment]
@anXieTyPB Any update on this? Again, from a look of it, it looks like there is connectivity issues between the nodes.

**Comment by anXieTyPB** [comment]
We were able to improve stability by setting more precise settings in our clients. Our main issues were caused by clients producing or consuming a lot of data in bulks in a very short time, causing our NATS server instance to not being able to handle them. 

We adjusted our configuration as follows to deal with this and are back to expected stability again running in a cluster.

- setMaxInFlight(10)

 Since a lot of messages were in flight at one point in time due to lots of requests in a short time, causing a lot of load on the nats cluster.

-  setManualAckMode(true) 

We automatically acknowledged all incoming messages which resulted in more messages being sent 
over time even though we did not process them yet. Changing this setting to true fixed this issue.

- MaxPubAcksInFlight : 20

We limited our Producers to a lower value since a lot of work is being done in parallel, resulting in a blocked NodeJS event loop for quite some time. When unblocking, a lot of messages get published at once that also have big payloads, which might result in the NATS streaming server instance being unable to deal with them

We keep testing our environment and report back if we can find another issue.

Regarding the Route connection created we can observe the following log messages in our development environment. It seems like a route is being created every minute. Is this expected behaviour?

```
[1] 2018/07/24 17:09:42.378039 [INF] IP1:6223 - rid:110636 - Route connection created
[1] 2018/07/24 17:09:42.399922 [INF] IP2:19744 - rid:110637 - Route connection created
[1] 2018/07/24 17:10:43.341925 [INF] IP2:19776 - rid:110638 - Route connection created
[1] 2018/07/24 17:10:43.397015 [INF] IP3:6223 - rid:110639 - Route connection created
[1] 2018/07/24 17:11:43.604188 [INF] IP2:19808 - rid:110640 - Route connection created
[1] 2018/07/24 17:11:43.612761 [INF] IP1:6223 - rid:110641 - Route connection created

```

**Comment by kozlovic** [comment]
If this dev env, I would recommend that you add the NATS debug property to figure out the route thing. This is the `-D` (notice the difference with `-SD` for streaming debug tracing) command line or, `debug: true` in the NATS specific config section. It would be interesting to see what happens with those routes.

**Comment by anXieTyPB** [comment]
We added the debug: true property to our nats configuration. Currently Route connection created does not show up as often and there seems to be a periodic ping timer every 2 minutes which appears to be a reasonable interval. After all, "Route connection created" is displayed only for a brief period of time and the cluster runs stable overall.

```
12:25:16 [1] 2018/08/08 12:25:16.598987 [DBG] IP1:49488 - rid:9 - Router Ping Timer
12:25:56 [1] 2018/08/08 12:25:56.696088 [DBG] IP2:50308 - rid:16 - Router Ping Timer
12:27:16 [1] 2018/08/08 12:27:16.599158 [DBG] IP1:49488 - rid:9 - Router Ping Timer
12:27:56 [1] 2018/08/08 12:27:56.696261 [DBG] IP2:50308 - rid:16 - Router Ping Timer
```

We still could not figure out the exact reason why the "Route connection created" issue happened on a local dev machine. We will keep observing it. Until then, I think the original issue has been resolved by applying above mentioned config changes to consumers/producers.

**Comment by kozlovic** [comment]
Thanks! Closing for now, will re-open if new information becomes available.

**Comment by snigdhasambitak** [comment]
Hello @anXieTyPB if possible could you please write a short document or guidelines on how did you set up an NATS Streaming Cluster in ECS. That would be really helpful to others. Specifications like the types of instances you used and service discovery patterns would be very helpful. Thanks!!

**Comment by caarlos0** [comment]
sorry to dig on an old issue... but I'm still able to reproduce this on 0.11.4, with small messages...

I made a reproducible here if anyone can take a look: https://github.com/caarlos0/nats-streaming-server-cluster/pull/1

on our production workload we are using java on clients and producers and get the same issue... several in a row most of the times...

**Comment by wallyqs** [comment]
thanks @caarlos0 for sharing how to reproduce that is helpful, we will take a look

**Comment by caarlos0** [comment]
np, its not very assertive though... but if you start multiple consumers/producers eventually it happens... messages are small (3 chars) on this test... on our production workload we have pretty big messages (few kb) and it happens too... not sure if msg size is relevant... 

if I would guess, it seems that at some point the leader is "too busy" or something and steps down for some reason. When this happens all clients disconnect (which I'm not sure if is expected or not).

restarting clients work again as well as changing ping configs to allow it to wait more time...

**Comment by caarlos0** [comment]
quick question: stan.java don't have the ping settings? (or maybe its just that I'm not being able to find them)

**Comment by caarlos0** [comment]
with nats in debug I see a lot of these too... 

```
node3_1  | [1] 2019/06/11 16:58:22.874301 [INF] STREAM: Deleting raft logs from 311093 to 339925
node3_1  | [1] 2019/06/11 16:58:22.886266 [INF] STREAM: Deletion took 11.9621ms
node1_1  | [1] 2019/06/11 16:58:36.880541 [DBG] 192.168.208.1:51016 - cid:16 - Client Ping Timer
node1_1  | [1] 2019/06/11 16:58:36.880714 [DBG] 192.168.208.1:51016 - cid:16 - Delaying PING due to activity 0s ago
node3_1  | [1] 2019/06/11 16:58:39.423645 [DBG] 192.168.208.1:50178 - cid:16 - Client Ping Timer
node3_1  | [1] 2019/06/11 16:58:39.423684 [DBG] 192.168.208.1:50178 - cid:16 - Delaying PING due to activity 0s ago
```

not sure how/if it helps

**Comment by kozlovic** [comment]
@caarlos0 Having a quick look at the repo you pointed to and I see a script that alternatively shuts down nodes. If that happens, then it is expected that there may be leader elections. Are you saying that the leader changes occur even without running this script?

**Comment by caarlos0** [comment]
> @caarlos0 Having a quick look at the repo you pointed to and I see a script that alternatively shuts down nodes. If that happens, then it is expected that there may be leader elections. Are you saying that the leader changes occur even without running this script?

yes, the script I was using for other tests, sorry for not clarifying it..

my guess (and its a wild one) is that the leader get too busy at some point and gives leadership up... I don't know how to verify that though... 

anyway, with that local setup there, running multiple consumers/producers eventually you'll get leader reelections... in our production environment it seems to happens more easily, maybe because message sizes are bigger overall or something...

**Comment by caarlos0** [comment]
Another pointer, regarding slowness: it seems waaaaaaaay slower on a cluster than it is on a single instance.

I expected it to be a little slower (like, say, 10% or so), but it is giving results like 5 times slower, even though the machines are the same size and networking is in internal VPC...

for reference, we have 3 x `n1-standard-4` with `500gb` of `pd-ssd` disk each (on boot disk), IOPS is always low (100 tops), networking low (30 bytes/sec tops), memory usage low (30% tops), CPU low (40% tops)... 

The slowness seems to be more on the producing side then the consuming... 

Is there anything we could do to improve this? I was thinking about increasing the machines sizes but they already seem oversized... any parameters I should at for tunning?

**Comment by kozlovic** [comment]
@caarlos0 As explained on Slack, it seems that you are producing very BIG messages (100MB) and NATS in general is not really meant to be used to single message that big (they could be chunked though). With streaming, you add persistence to the mix, and in clustering mode, the message is replicated (actually persisted in RAFT log for replication and streaming channel message log) so not really the best fit. You have observed slow consumer for routes between nodes, which will cause leader election.

**Comment by caarlos0** [comment]
yeah that makes sense, will try to change things here to lower these limits...

SQS is 1256kb by default, is this a good value for NATS too?

**Comment by kozlovic** [comment]
The default max payload in NATS (as you probably know) is 1MB, and in case of NATS Streaming, it is actually a lower payload due to the protobuf overhead. Again, as describe before, big messages are not the best fit with NATS, especially if there are lots of matching subscriptions.

That being said, even with small messages, you have been showing that there are cases where the traffic is enough to cause leader re-election, which is caused by nodes not being able to contact their peers (RAFT heartbeats). The thing I don't like is that it causes the client PINGs to fail and this is because while the leader steps down, it unsubscribes to the client PINGs requests and since the re-election takes more than 2 seconds, and your test case uses Pings(1, 3), the application is likely to fail before the new leader is elected. You could set it to Pings(1, 10) and likely the applications would not fail with the PINGs error. This is not the root of the issue, the issue is that the RAFT heartbeats get delayed by the replication traffic and unfortunately I am not sure I can solve this problem. You can however increase the RAFT heartbeat timeout to minimize the risk of that re-election. To increase the timeout to 2s, you would set `raft_heartbeat_timeout: "2s"` for instance.

**Comment by caarlos0** [comment]
got it! thanks again for the explanation, it's very helpful.

we ended up going to the fault-tolerant model with google's managed nfs (glusterfs), still testing but so far performance looks way better.

will work on messages size limits as well... thanks!

**Comment by caarlos0** [comment]
got it! thanks again for the explanation, it's very helpful.

we ended up going to the fault-tolerant model with google's managed nfs (glusterfs), still testing but so far performance looks way better.

will work on messages size limits as well... thanks!
